src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,334 @@
+                        });
+                        break;
+                    default:
+                        ExceptionUtilities.UnexpectedValue(sortOption);
+                        break;
+                }
+
+                return sortedDocumentSymbols.ToImmutable();
+            }
+        }
+
+        /// <summary>
+        /// Returns an immutable array of DocumentSymbolData such that each node matches the given pattern.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolData> SearchDocumentSymbolData(
+            ImmutableArray<DocumentSymbolData> documentSymbolData,
+            string pattern,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var _ = ArrayBuilder<DocumentSymbolData>.GetInstance(out var filteredDocumentSymbols);
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
+
+            foreach (var documentSymbol in documentSymbolData)
+            {
+                var filteredChildren = SearchDocumentSymbolData(documentSymbol.Children, pattern, cancellationToken);
+                if (SearchNodeTree(documentSymbol, patternMatcher, cancellationToken))
+                    filteredDocumentSymbols.Add(new DocumentSymbolData(documentSymbol, filteredChildren));
+            }
+
+            return filteredDocumentSymbols.ToImmutable();
+
+            // Returns true if the name of one of the tree nodes results in a pattern match.
+            static bool SearchNodeTree(DocumentSymbolData tree, PatternMatcher patternMatcher, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (patternMatcher.Matches(tree.Name))
+                    return true;
+
+                foreach (var childItem in tree.Children)
+                {
+                    if (SearchNodeTree(childItem, patternMatcher, cancellationToken))
+                        return true;
+                }
+
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Converts an immutable array of DocumentSymbolData to an immutable array of DocumentSymbolUIItems.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolUIItem> GetDocumentSymbolUIItems(ImmutableArray<DocumentSymbolData> documentSymbolData, IThreadingContext threadingContext)
+        {
+            using var _ = ArrayBuilder<DocumentSymbolUIItem>.GetInstance(out var documentSymbolItems);
+            foreach (var documentSymbol in documentSymbolData)
+            {
+                var children = documentSymbol.Children.IsEmpty ? ImmutableArray<DocumentSymbolUIItem>.Empty : GetDocumentSymbolUIItems(documentSymbol.Children, threadingContext);
+                var documentSymbolItem = new DocumentSymbolUIItem(documentSymbol, children, threadingContext);
```

Comment:
Nit: 

```suggestion
                var children = GetDocumentSymbolUIItems(documentSymbol.Children, threadingContext);
                var documentSymbolItem = new DocumentSymbolUIItem(documentSymbol, children, threadingContext);
```

GetDocumentSymbolUIItems will already properly handle an empty list of children, so no need to special case here.